File: src/AI/Rystem.PlayFramework/Domain/Models/AiSceneResponse.cs
namespace Rystem.PlayFramework;

/// <summary>
/// Represents a response from a scene execution.
/// </summary>
public sealed class AiSceneResponse
{
    /// <summary>
    /// Current status of the execution.
    /// </summary>
    public AiResponseStatus Status { get; set; }

    /// <summary>
    /// Scene name (if applicable).
    /// </summary>
    public string? SceneName { get; set; }

    /// <summary>
    /// Function/tool name that was called (if applicable).
    /// </summary>
    public string? FunctionName { get; set; }

    /// <summary>
    /// Function/tool arguments (if applicable).
    /// </summary>
    public string? FunctionArguments { get; set; }

    /// <summary>
    /// Response message from AI or tool execution.
    /// </summary>
    public string? Message { get; set; }

    /// <summary>
    /// Error message (if Status is Error).
    /// </summary>
    public string? ErrorMessage { get; set; }

    /// <summary>
    /// Input tokens used in this operation.
    /// </summary>
    public int? InputTokens { get; set; }

    /// <summary>
    /// Cached input tokens (10% cost).
    /// </summary>
    public int? CachedInputTokens { get; set; }

    /// <summary>
    /// Output tokens generated in this operation.
    /// </summary>
    public int? OutputTokens { get; set; }

    /// <summary>
    /// Total tokens (input + cached + output).
    /// </summary>
    public int? TotalTokens { get; set; }

    /// <summary>
    /// Cost of this specific operation (null if no LLM call).
    /// </summary>
    public decimal? Cost { get; set; }

    /// <summary>
    /// Cumulative cost so far in the execution.
    /// </summary>
    public decimal TotalCost { get; set; }

    /// <summary>
    /// Timestamp of this response.
    /// </summary>
    public DateTimeOffset Timestamp { get; set; } = DateTimeOffset.UtcNow;

    /// <summary>
    /// Additional metadata.
    /// </summary>
    public Dictionary<string, object> Metadata { get; set; } = [];
}

File: src/AI/Rystem.PlayFramework.Client/src/rystem/src/models/AiSceneResponse.ts
/**
 * Response status from PlayFramework execution.
 */
export type AiResponseStatus = "Planning" | "Running" | "Error" | "Completed";

/**
 * Single step response from PlayFramework (used in step-by-step streaming).
 */
export interface AiSceneResponse {
    /**
     * Execution status.
     */
    status: AiResponseStatus;

    /**
     * Response message/text.
     */
    message?: string;

    /**
     * Error message (if status==="Error").
     */
    errorMessage?: string;

    /**
     * Total tokens used so far.
     */
    totalTokens?: number;

    /**
     * Total cost so far.
     */
    totalCost?: number;

    /**
     * Metadata from this step.
     */
    metadata?: Record<string, any>;

    /**
     * Tool calls executed in this step.
     */
    toolCalls?: any[];
}

/**
 * Completion marker for SSE streaming.
 */
export interface CompletionMarker {
    status: "completed";
}

/**
 * Error marker for SSE streaming.
 */
export interface ErrorMarker {
    status: "error";
    errorMessage: string;
}

/**
 * SSE event types.
 */
export type SSEEvent = AiSceneResponse | CompletionMarker | ErrorMarker;
